### PR TITLE
Add com.vcvrack.Rack2

### PR DESCRIPTION
**Pull request by zhabba**
### Please confirm your submission meets all the criteria

- [x] Please describe the application briefly.
  VCV Rack 2 is a powerful, modular synthesizer software that emulates Eurorack hardware on your computer.
- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:** 

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: com.vcvrack.Rack2.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  - --filesystem=host
+  - --filesystem=xdg-run/pipewire-0
+  - --env=XDG_CURRENT_DESKTOP=GNOME
+
+x-gl-version: &gl-version '1.4'
+x-gl-versions: &gl-versions 23.08;23.08-extra;1.4
+x-gl-merge-dirs: &gl-merge-dirs vulkan/icd.d;glvnd/egl_vendor.d;lib/dri;lib/d3d;vulkan/explicit_layer.d;vulkan/implicit_layer.d
+
+add-extensions:
+  org.freedesktop.Platform.GL:
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:24]: why is this necessary?

File: com.vcvrack.Rack2.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    subdirectories: true
+    no-autodownload: true
+    autodelete: false
+    add-ld-path: lib
+    merge-dirs: *gl-merge-dirs
+    download-if: active-gl-driver
+    enable-if: active-gl-driver
+    autoprune-unless: active-gl-driver
+
+  org.freedesktop.Platform.VAAPI.Intel:
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:37]: and this?

File: com.vcvrack.Rack2.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        path: com.vcvrack.Rack2.desktop
+      - type: file
+        path: metainfo.xml
+      - type: script
+        dest-filename: rack_wrapper.sh
+        commands:
+          - cd /app
+          - exec /app/Rack --system /app --user $XDG_DATA_HOME $@
+      - type: archive
+        url: https://vcvrack.com/downloads/RackFree-2.5.2-lin-x64.zip
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:72]: why not building from source?

File: metainfo.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  <url type="homepage">https://vcvrack.com/</url>
+  <url type="bugtracker">https://github.com/VCVRack</url>
+  <url type="help">https://vcvrack.com/support</url>
+  <developer id="com.vcvrack">
+    <name>VCV</name>
+  </developer>
+  <update_contact>zhabba108_at_gmail.com</update_contact>
+  <releases>
+    <release version="2.5.2" date="2024-05-09">
+      <description>Virtual Eurorack Studio</description>
```
Comment (hfiguiere) [inline metainfo.xml:37]: this is supposed to be release notes.

File: metainfo.xml
Comment (hfiguiere) [inline metainfo.xml:1]: this should be submitted upstream.

File: com.vcvrack.Rack2.desktop
Comment (hfiguiere) [inline com.vcvrack.Rack2.desktop:1]: this belong upstream and should be submitted there if it doesn't exist.

File: com.vcvrack.Rack2.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+runtime: org.freedesktop.Platform
+runtime-version: &runtime-version '23.08'
+sdk: org.freedesktop.Sdk
+command: rack
+
+finish-args:
+  - --share=ipc
+  - --share=network
+  - --socket=wayland
+  - --socket=x11
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:11]: it's either x11 alone or wayland + fallback-x11. Not both.

File: com.vcvrack.Rack2.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  - --socket=pulseaudio
+  - --device=all
+  - --allow=bluetooth
+  - --filesystem=host
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:15]: this is a very broad filesystem permission

File: com.vcvrack.Rack2.yaml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  - --filesystem=xdg-run/pipewire-0
+  - --env=XDG_CURRENT_DESKTOP=GNOME
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:17]: why setting this?

File: com.vcvrack.Rack2.yaml (hunk 1 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:24]: It needed because of application is built with GLFW library - OpenGL is required for app to function properly.

File: com.vcvrack.Rack2.yaml (hunk 2 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:37]: Same here - OpenGL in case of Intel GPU

File: com.vcvrack.Rack2.yaml (hunk 3 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:72]: It's to much for me not being C++ developer. I could build it in standard environment but unlikely in flatpak env.
It's a huge dependencies list: https://github.com/VCVRack/Rack

File: metainfo.xml (hunk 4 above)
Comment (zhabba) [reply metainfo.xml:37]: Make sense, will fix.

File: com.vcvrack.Rack2.yaml (hunk 1 above)
Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml:24]: so you don't need it. The question was rhetorical.

File: com.vcvrack.Rack2.yaml (hunk 2 above)
Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml:37]: same. you don't. Not sure what erroneous documentation tell to do so but it's not correct.

File: com.vcvrack.Rack2.yaml (hunk 3 above)
Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml:72]: it's not about C++.

File: com.vcvrack.Rack2.yaml (hunk 7 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:17]: -filesystem=xdg-run/pipewire-0  as I learned from other repos is for Jack server support, app can be configured to use it.
--env=XDG_CURRENT_DESKTOP=GNOME - might be a leftover from previous experiments, will remove it.

File: com.vcvrack.Rack2.yaml (hunk 5 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:11]: Never worked for me with fallback-x11 option. Will remove x11 option.

File: com.vcvrack.Rack2.yaml (hunk 6 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:15]: Well, I saw similar in many repos, just copy-pasted.

Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml:15]: two wrongs do not make a right.

File: metainfo.xml
Comment (zhabba) [reply metainfo.xml:1]: Upstream is not accepting contributions.

File: com.vcvrack.Rack2.yaml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    version: *gl-version
+    versions: *gl-versions
+    subdirectories: true
+    no-autodownload: true
+    autodelete: false
+    add-ld-path: lib
+    merge-dirs: *gl-merge-dirs
+    download-if: active-gl-driver
+    enable-if: active-gl-driver
+    autoprune-unless: active-gl-driver
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:35]:
```suggestion
```

File: com.vcvrack.Rack2.yaml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    version: *runtime-version
+    versions: *runtime-version
+    subdirectories: true
+    no-autodownload: true
+    autodelete: false
+    add-ld-path: lib
+    merge-dirs: *gl-merge-dirs
+    download-if: have-intel-gpu
+    enable-if: have-intel-gpu
+    autoprune-unless: have-intel-gpu
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:48]:
```suggestion
```

File: com.vcvrack.Rack2.yaml [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+modules:
+  - name: rack2free
+    buildsystem: simple
+    build-commands:
+      - mkdir -p ${FLATPAK_DEST}/lib/GL
+      - mkdir -p ${FLATPAK_DEST}/dri/intel-vaapi-driver
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:55]:
```suggestion
```

File: com.vcvrack.Rack2.yaml (hunk 7 above)
Comment (bbhtt) [reply com.vcvrack.Rack2.yaml:17]: > -filesystem=xdg-run/pipewire-0

It's only needed if the app is using Pipewire apis directly

File: com.vcvrack.Rack2.yaml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+command: rack
+
+finish-args:
+  - --share=ipc
+  - --share=network
+  - --socket=wayland
+  - --socket=x11
+  - --socket=pulseaudio
+  - --device=all
+  - --allow=bluetooth
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:14]: why is `all` devices and bluetooth needed ?

File: com.vcvrack.Rack2.yaml (hunk 1 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: I've reworked manifest accordingly, double-checked, fixed.

File: com.vcvrack.Rack2.yaml (hunk 3 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:34]: I would insist on using official distribution of the app as a base.

File: metainfo.xml (hunk 4 above)
Comment (zhabba) [reply metainfo.xml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 5 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 7 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 8 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 10 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 11 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 9 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 6 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 2 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Fixed

File: com.vcvrack.Rack2.yaml (hunk 7 above)
Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml]: > > -filesystem=xdg-run/pipewire-0
> 
> It's only needed if the app is using Pipewire apis directly

It's needed for JACK as it is the only way JACK works: pw-jack

File: com.vcvrack.Rack2.yaml (hunk 3 above)
Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml:34]: I would insist on building from source.

File: com.vcvrack.Rack2.yaml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    sources:
+      - type: file
+        path: com.vcvrack.Rack2.desktop
+      - type: file
+        path: metainfo.xml
+      - type: script
+        dest-filename: rack_wrapper.sh
+        commands:
+          - cd /app
+          - exec /app/Rack --system /app --user ${XDG_DATA_HOME} $@
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:32]:
```suggestion
          - exec /app/Rack --system /app --user ${XDG_DATA_HOME} "$@"
```

Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml:32]: also the exectuable `/app/Rack` is in the wrong location

File: com.vcvrack.Rack2.yaml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      - type: archive
+        url: https://vcvrack.com/downloads/RackFree-2.5.2-lin-x64.zip
+        sha256: 6c7bba5f3ccf8fe667771ddd2619c6a17f84b5dfd31d3683f56c253e27629ca1
+
+cleanup:
+  - rack_wrapper.sh
```
Comment (hfiguiere) [inline com.vcvrack.Rack2.yaml:38]: this doesn't do what you think it does.

```suggestion
```

File: metainfo.xml
Comment (hfiguiere) [reply metainfo.xml:1]: haviing the file properly named would be nice too.

File: com.vcvrack.Rack2.desktop
Comment (zhabba) [reply com.vcvrack.Rack2.desktop:1]: Same here: https://github.com/VCVRack/Rack#contributions

File: metainfo.xml [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    <p>For more details, visit VCV Rack's official site.</p>
+    <p>This wrapper is not verified by, affiliated with, or supported by VCV</p>
+  </description>
+  <categories>
+    <category>Audio</category>
+    <category>Music</category>
+  </categories>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://vcvrack.com/Rack2/plugin.webp</image>
```
Comment (hfiguiere) [inline metainfo.xml:24]: webp is NOT a supported format.

Comment (bbhtt) [reply metainfo.xml:24]: It works now, we added webp-pixbuf-loader to org.flatpak.Builder.

Comment (hfiguiere) [reply metainfo.xml:24]: nevermind then.

File: com.vcvrack.Rack2.yaml (hunk 12 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:32]: > also the exectuable `/app/Rack` is in the wrong location

Executable is exactly where it intended by developers to be: https://vcvrack.com/manual/Installing#Installing-on-Linux. To call it I'm using wrapper script.

File: com.vcvrack.Rack2.yaml (hunk 13 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:38]: It does :) It deletes `rack_wrapper.sh` once it's installed in `/app/bin`  exactly as described in [doc](https://docs.flatpak.org/en/latest/manifests.html), cleanup section.
I'd maybe move it into the module scope but keep it.

Comment (hfiguiere) [reply com.vcvrack.Rack2.yaml:38]: that because to should never reach there then. cleanup is to remove things installed by the build system (there is none) that are not wished for version reasons.

File: metainfo.xml (hunk 14 above)
Comment (zhabba) [reply metainfo.xml:24]: Heck, I need to revert changes then :)

File: com.vcvrack.Rack2.yaml (hunk 13 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: I'm confused: if I'm removing `cleanup` section and run `flatpak  run --command=sh  com.vcvrack.Rack2` I see the `rack_wrapper.sh` in the /app dir. If I keep `cleanup` section - file been removed.

File: com.vcvrack.Rack2.yaml (hunk 3 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:34]: WIP

File: com.vcvrack.Rack2.yaml (hunk 13 above)
Comment (bbhtt) [reply com.vcvrack.Rack2.yaml]: It's coming in `/app` because `cp` with `*` is copying it. If you move it before copying, it won't be needed.

File: com.vcvrack.Rack2.yaml (hunk 3 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml]: Manifest was reworked to build everything from sources.

File: com.vcvrack.Rack2.yaml [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    sources:
+    - type: git
+      url: https://github.com/fletcher/MultiMarkdown-6.git
+      tag: 6.7.0
+
+  - name: rack2free
+    buildsystem: simple
+    build-options:
+      build-args:
+        - --share=network
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:28]: build-args has no effect on Flathub, Build is done offline. https://docs.flathub.org/docs/for-app-authors/requirements#no-network-access-during-build

File: com.vcvrack.Rack2.desktop
Comment (bbhtt) [inline com.vcvrack.Rack2.desktop:1]: Desktop file should be moved to upstream as well

File: com.vcvrack.Rack2.yaml [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    build-commands:
+      - |
+        set -e
+        J_CPUS=$(($(getconf _NPROCESSORS_ONLN)+1))
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:32]:
```suggestion
```

File: com.vcvrack.Rack2.yaml [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        rm -rf plugins/.gitignore
+        make -j${J_CPUS} dep
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:34]: Use `$FLATPAK_BUILDER_N_JOBS`

File: com.vcvrack.Rack2.yaml [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        install -Dm644 metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
+        install -Dm755 ${FLATPAK_ID}.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+        install -Dm644  res/icon.png ${FLATPAK_DEST}/share/icons/hicolor/256x256/apps/com.vcvrack.Rack2.png
+        install -Dm755 rack_wrapper.sh ${FLATPAK_DEST}/bin/rack
+        rm -rf ${FLATPAK_DEST}/bin/m* ${FLATPAK_DEST}/lib/ ${FLATPAK_DEST}/share/doc/ ${FLATPAK_DEST}/share/texmf/
+    sources:
+      - type: git
+        url: https://github.com/VCVRack/Rack.git
+        commit: 58f2482df860fe878581423aedc8cadb1d94eb50
+        dest: .
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:50]:
```suggestion
```

File: com.vcvrack.Rack2.desktop
Comment (bbhtt) [reply com.vcvrack.Rack2.desktop:1]: Ah nevermind I forgot you said upstream is not accepting these.

File: com.vcvrack.Rack2.yaml [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        pushd plugins/Fundamental
+        make -j${J_CPUS} dist
+        popd
+        make -j${J_CPUS} dist
+        cp -R dist/Rack2Free/* ${FLATPAK_DEST}/
+        install -Dm644 metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
+        install -Dm755 ${FLATPAK_ID}.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+        install -Dm644  res/icon.png ${FLATPAK_DEST}/share/icons/hicolor/256x256/apps/com.vcvrack.Rack2.png
+        install -Dm755 rack_wrapper.sh ${FLATPAK_DEST}/bin/rack
+        rm -rf ${FLATPAK_DEST}/bin/m* ${FLATPAK_DEST}/lib/ ${FLATPAK_DEST}/share/doc/ ${FLATPAK_DEST}/share/texmf/
```
Comment (bbhtt) [inline com.vcvrack.Rack2.yaml:45]: Use `cleanup` also why remove `/app/lib` ?

Comment (bbhtt) [reply com.vcvrack.Rack2.yaml:45]: https://docs.flatpak.org/en/latest/manifests.html#cleanup

File: metainfo.xml
Comment (hfiguiere) [reply metainfo.xml:1]: This last thing wasn't done. The proper name would be the one you install it under.

File: com.vcvrack.Rack2.yaml (hunk 15 above)
Comment (zhabba) [reply com.vcvrack.Rack2.yaml:28]: Oh hell, I'm giving up here as I don't see any reason to reinvent wheels and mimic all Makefiles recursively. That's why I was insisting on using prebuilt package from developer's site. I'm closing PR and happily use my precious local bundle.
